tracing/src/main/java/com/palantir/remoting2/tracing/Tracers.java

```diff
@@ -96,11 +100,14 @@ public static Runnable wrap(Runnable delegate) {
         @Override
         public V call() throws Exception {
             Trace originalTrace = Tracer.copyTrace();
+            String originalMdcTraceIdValue = MDC.get(MDC_KEY);
             Tracer.setTrace(trace);
+            MDC.put(MDC_KEY, trace.getTraceId());
```

Comment:
Any fix would have to be piecemeal right? In all those cases we are going to need:
- A thread local strucuture to hold the information we care about (traceId, userId...)
- A wrapper for ExecutorServices to propagate those thread local structures to their threads

We are just saying that this thread local structure, instead of being the MDC, will be a custom thing for every case. I can see the point of a strongly typed structure being better than the string-keyed MDC though. If you think the logback extension thing is smooth (it would need to depend on all the libraries that produce those thread locals), we could go for that. Otherwise, I don't think it makes the pain much smaller